### ISSUE TITLE
Openshift support

### ISSUE THREAD

**Issue by sreekanth-ragi** [AjayTripathy/kubemetrics-quickstart#3]
Hi Team,
We intend to try this out on Openshift stack, is it possible?

Thanks,
Sreekanth

**Comment by dwbrown2** [comment]
Hi Sreekanth, we have a couple of teams running Kubecost on Openshift so you should be good. We highly encourage people to use the instructions on this project now, specifically our Helm chart there: 

https://github.com/kubecost/cost-model

Feel free to join our [Slack](https://join.slack.com/t/kubecost/shared_invite/enQtNTA2MjQ1NDUyODE5LTg0MzYyMDIzN2E4M2M5OTE3NjdmODJlNzBjZGY1NjQ3MThlODVjMGY3NWZlNjQ5NjIwNDc2NGU3MWNiM2E5Mjc) if you have immediate questions. Let us know if you have any challenges.

**Comment by sreekanth-ragi** [comment]
Hi 

Thanks for the answer, I was able to deploy the cost-model on Openshift, but the pod does not come up with the following error.

F0503 09:46:27.109402       1 main.go:173] Failed to use Prometheus at http://prometheus-k8s.openshift-monitoring.svc:9091  Error: Get http://prometheus-k8s.openshift-monitoring.svc:9091/api/v1/status/config:  net/http: HTTP/1.x transport connection broken: malformed HTTP response "\x15\x03\x01\x00\x02\x02"

The above is a service URL, I also tried the route URL which is https and I get the following error.

https://prometheus-k8s-openshift-monitoring.infra.saasitc-gce-shared.ragsr01.cs.saas.ca.com/

Probably I'm doing something wrong. Any pointers that I could look into.

Thanks,
Sreekanth

**Comment by dwbrown2** [comment]
Hi Sreekanth, happy to help. Can you confirm that you are trying to deploy the `cost-model` pod directly, as opposed to the Kubecost Helm chart from kubecost.com?   

I suspect that your Prometheus service address is incorrect. You should able to provide it in this format without the port included: `http://<service-name>.<namespace-name>.svc.cluster.local`

**Comment by sreekanth-ragi** [comment]
Yes I’m trying the cost model pod directly, I think I tried without the port number in the service URL, but guess that dint work either. I will re check and post the results of what I see

Thanks

**Comment by dwbrown2** [comment]
Sounds goods, I would start by confirming that the Prometheus server is indeed available at the fqdn that you provide. Not sure about your exact DNS setup but this article covers fqdn construction: https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/

**Comment by srinivma1** [comment]
I am also facing same issue while deploying cost-model in Openshift. Any luck? Were you able to solve the issue?

**Comment by dwbrown2** [comment]
It's worth noting that srinivma1's Prometheus installation is using a re-encryption route. I do not believe this is supported by kubecost today. 

https://docs.openshift.com/container-platform/3.9/architecture/networking/routes.html

**Comment by srinivma1** [comment]
Thanks for the answer. I am requoting comment posted by Sreekanth.

I was able to deploy the cost-model on Openshift, but the pod does not come up with the following error. Tried both **service and route approach** to connect to existing prometheus server.

**F0503 09:46:27.109402 1 main.go:173] Failed to use Prometheus at http://prometheus-k8s.openshift-monitoring.svc Error: Get http://prometheus-k8s.openshift-monitoring.svc/api/v1/status/config: net/http: HTTP/1.x transport connection broken: malformed HTTP response "\x15\x03\x01\x00\x02\x02"**

The above is a service URL, I also tried the route URL(re-encryption route) which is https and I get the following error.

https://prometheus-k8s-openshift-monitoring.infra.saasitc-gce-shared.ragsr01.cs.saas.ca.com/

**Error: Certificate signed by unknown authority**

I was atleast expecting the connection should work with service URL (not the route URL) but even that is not working.

Could you please try yourself on an Openshift cluster and let me know if you are able to reproduce the issue or not?

**Comment by sreekanth-ragi** [comment]
Hi Webb,

I tried using the service URL with http and https and I see error in both the conditions.

Failed to use Prometheus at http://prometheus-k8s.openshift-monitoring.svc  Error: Get http://prometheus-k8s.openshift-monitoring.svc/api/v1/status/config:  dial tcp 172.30.85.171:80: i/o timeout

Failed to use Prometheus at https://prometheus-k8s.openshift-monitoring.svc  Error: Get https://prometheus-k8s.openshift-monitoring.svc/api/v1/status/config:  dial tcp 172.30.85.171:443: i/o timeout

The prometheus URL is working and the accessible over the route.

Thanks,
Sreekanth

**Comment by dwbrown2** [comment]
Closing now that discussion is moving to https://github.com/kubecost/cost-model/issues/74

**Comment by AjayTripathy** [comment]
@sreekanth-ragi This seems like an authentication error based on [this issue](https://github.com/kubecost/cost-model/issues/74) . Can you confirm if your openshift installation requires authentication for intra-cluster communication?

Can you curl your prometheus instance from any other container in your cluster with / without any auth tokens to see if the request succeeds?

**Comment by dwbrown2** [comment]
@sreekanth-ragi, @srinivma1 was able to get this working by doing the following. I have not reviewed close, but I thought this might be helpful for you. Mehesh, let me know if there are other changes made. 

Added these commands:

oc adm policy add-scc-to-group anyuid system:authenticated 
oc adm policy add-scc-to-user anyuid -z kubecost-cost-analyzer -n kubecost
oc adm policy add-scc-to-user anyuid -z default -n kubecost
oc adm policy add-scc-to-user anyuid -z kubecost-grafana -n kubecost (edited) 

Prometheus pod on startup was giving permission denied error on /data/lock file
Had to add '--storage.tsdb.no-lockfile' in the container args section .

Also, added following under securityContext in prometheus deployment.

securityContext:
     fsGroup: 2000
     runAsNonRoot: true
     runAsUser: 1000

**Comment by sreekanth-ragi** [comment]
@dwbrown2 - apologies for the delayed response, as indicated by you I'm setting this up from the following link 
https://github.com/kubecost/cost-model
Does the above commands still hold good or do you want me to set this up using 
https://kubecost.com/install?ref=home

Thanks for all your support and time.

Sreekanth

**Comment by dwbrown2** [comment]
Hi @sreekanth-ragi, either option still works! Installing from kubecost.com will give you our UI and more functionality. Both are working for teams on Openshift today. 

Check out comments here: https://github.com/kubecost/cost-model/issues/74 and let me know if you have any other questions!
